### ISSUE TITLE
Discovering container assignment to the GPUs

### ISSUE THREAD

**Issue by nadgowdas** [cloudviz/agentless-system-crawler#334]
## Description
For GPU metric collection, we run crawler on host and collect metric in the form:

```
metadata    "metadata"    {"system_type":"host","timestamp":"2017-08-09T21:50:49+0000","namespace":"10.184.164.164","features":"os,gpu","uuid":"c0e9c1cf-7370-4825-b48e-e858f4a3fd2e"}
gpu    "gpu0"    {"memory":{"total":12205,"used":0,"free":12205},"temperature":31,"power":{"draw":27,"limit":149},"utilization":{"gpu":0,"memory":0}}
gpu    "gpu1"    {"memory":{"total":12205,"used":0,"free":12205},"temperature":28,"power":{"draw":29,"limit":149},"utilization":{"gpu":0,"memory":0}}
```

For DLaaS and in general, the requirement is to discover the containers which are using respective GPUs and add that information as part of metric collection.

Assumptions:
---------
1) GPUs are not shared between containers simultaneously.
2) Container is allocated 1 or more GPUs on node

@canturkisci @sahilsuneja1 @fabolive

**Comment by fabolive** [comment]
@nadgowdas and I generated load on the GPUs of the DLaaS dev cluster, and validated that we can identify the pids of the processes using the GPUs. From that point on, we should be able to map the host processes to the containers/pods.

**Comment by nadgowdas** [comment]
Updates:
-----------
1. @sahilsuneja1 has implemented the change in GPU plugin 
2. @fabolive  and me, we made couple of fixes in the plugin and tested it on DLaaS cluster. 
3. Our earlier premise was: we get the PID of the process using GPU, and then match that with PID of all local containers. It assumed, that the entrypoint of the container is the same process using the GPU.
4. But, this hypothesis did not work in practice. We discovered from DLaaS jobs, that the entrypoint is usually a script that triggers another process that uses GPU.

Options:
----------
1. In Host-crawler collecting GPU metric, we just collect the PID information for process using GPU and not associate it with any container.
In Container-crawler, we crawl with `--features=process` and get process information inside container. 
Since, we crawl with shared `pid` namespace, these PIDs will match.
But, this is out-of-band GPU->container association that need to be done and it won't be available in logmet.

2. In GPU plugin, we implement brute-force, to crawl each container for all processes and match them.

Next on:
----------
1. @fabolive will check with Priya to understand use-case more clearly and discuss current limitation. Then, we will decide on changes to plugin plan.

**Comment by nadgowdas** [comment]
Update:
----------
1. @sahilsuneja1 has implemented a change to associate PID using GPU to the container name.
2. @fabolive and we tested the change on DLaaS cluster and its working. 
3. We will soon merge this change into master branch.

**Comment by fabolive** [comment]
I verified that the mtgraphite emitter works as expected with the legacy Logmet endpoint  `/metrics.opvis.bluemix.net:9095`, but it does NOT seem to work with the new endpoint `/metrics.ng.bluemix.net:9095`. I believe the legacy endpoint will be supported until the middle of October. We must keep that in mind and check why the emitter does not work with the new endpoint.

I have created a Grafana dashboard to plot network and disk bandwidth for the hosts of the DLaaS development cluster. I will add GPU data once the latest changes to the crawler are merged into the mainstream code.

cc: @canturkisci, @sahilsuneja1, @nadgowdas
